test/Microsoft.ML.Tests/LearningRateSchedulerTest.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.ML.RunTests;
+using Microsoft.ML.Transforms;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Tests
+{
+    public sealed class LearningRateSchedulerTest : TestDataPipeBase
```

Comment:
>LearningRateSchedulerTest [](start = 24, length = 25)

Can we also put a test with a ImageClassification API? I know it will increase the build time but I will prefer we have that test.

Reply:
Actually I disagree there is already a test that uses a LR scheduler which tests that the classification API can work with an LR scheduler we should not be duplicating that test to test with this scheduler, instead this is a much more correct unit test

---
In reply to: [339268046](https://github.com/dotnet/machinelearning/pull/4370#discussion_r339268046) [](ancestors = 339268046)

Reply:
Your code passes a learning rate to Image Classification API that is passed to TF graph runner and by adding an image classification test you would have added an e2e test. I appreciate your isolated test but I will prefer we have a unit-test that also test image classification API with this LR since our users will also be using it this way.

---
In reply to: [339268678](https://github.com/dotnet/machinelearning/pull/4370#discussion_r339268678) [](ancestors = 339268678,339268046)

Reply:
[https://github.com/dotnet/machinelearning/blob/b1788c36268999e66df00356417e604f08d2a568/test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs#L1485](https://github.com/dotnet/machinelearning/blob/b1788c36268999e66df00356417e604f08d2a568/test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs#L1485)

Already tests that a learning rate when passed to the image classification API is properly handled by the API, we should not be running the same test with a different LR scheduler that is not a useful test instead we need to confirm that this scheduler is producing the results we expect

---
In reply to: [339271171](https://github.com/dotnet/machinelearning/pull/4370#discussion_r339271171) [](ancestors = 339271171,339268678,339268046)

Reply:
@eerhardt to comment further. My feelings: https://cdn-images-1.medium.com/max/1600/1*UtZzMT32fRMnSN-HmgiSVQ.gif

Reply:
My opinion is that both levels of testing are important, and there are tradeoffs on which level you write your tests.

1. If all of your tests are unit tests, you can't tell if the components work when used together.
2. If all of your tests are functional tests, it is a lot harder to test different permutations of inputs/settings. Both trying to set all those settings set at the functional level, and in the amount of time it is going to take to execute end-to-end tests.

Having a good balance is crucial. It ensures that a component works correctly with all different inputs, but our tests don't become so long to run that they are no longer maintainable.

In this case, you can view the whole `LearningRateScheduler` system as the "component" or each individual derived class as the "component". This is why I think you two seem to be having a disagreement. It appears @codemzs views each individual derived class as the "component" that should have a functional test to ensure it works with the rest of the system. And @bpstark views the whole `LearningRateScheduler` system as the "component" that already does have a functional test, and so there is no need testing each individual subclass.

If I were doing this myself, I would see how long the existing functional test takes. If it takes so long that doubling the time would be unacceptable, I wouldn't add a new functional test. But if the existing functional test is fast enough that adding a second one wouldn't be noticeable, I would add it.